Optimize Min/Max using Delta metadata

felipepessoto opened the pull request:
## Description

Follow up of https://github.com/delta-io/delta/issues/1192, which optimizes COUNT. This PR adds support for MIN/MAX as well.

Fix #2092

## How was this patch tested?

Created additional unit tests to cover MIN/MAX.

## Does this PR introduce _any_ user-facing changes?

Only performance improvement

felipepessoto commented:
Yes, I made these changes while the SELECT Count was in review, I think I can refine this.

felipepessoto commented:
@scottsand-db it is ready to review. Thanks

vkorukanti commented on core/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -17,53 +17,297 @@
+  def extractGlobalColumnStats(tahoeLogFileIndex: TahoeLogFileIndex):
+  CaseInsensitiveMap[DeltaColumnStat] = {
+
+    // TODO Update this to work with DV (https://github.com/delta-io/delta/issues/1485)
```
Should we handle the DV support? The read support is already there. Check out the `DeletionVectorsSuite`

felipepessoto replied:
I need to think more about how it will work. I added distinctCount (for partition values only) and nullCount just for completeness, it is probably better to remove them.

For min and max, I need to check if the value found are from a AddFile.stats where tightBounds=true.

felipepessoto replied:
It seems the implementation to update the stats tightBounds is still not done. We need it to create the unit tests.

felipepessoto replied:
Should we wait for it? Or could we proceed and address DV later?

vkorukanti replied:
We can add support later, but at least we shouldn't proceed with the optimization if DVs are preset. Otherwise it will give wrong results.

felipepessoto replied:
Done.
We have ETA when DV will be released?

felipepessoto commented on core/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -17,53 +17,291 @@
     plan.transformUpWithSubqueries {
-      case agg@CountStarDeltaTable(countValue) =>
-        LocalRelation(agg.output, Seq(InternalRow(countValue)))
+      case agg@AggregateDeltaTable(tahoeLogFileIndex) =>
```
Probably not. Show() adds CAST to the columns. I need to find a clean solution to handle it, this new implementation is more complex because it doesn't assume a specific query format like previous one. We can have SELECT COUNT(*), CAST(COUNT(*) AS String) FROM TABLE.

felipepessoto replied:
I added a special case for COUNT + show(). I hope this helps for now.

felipepessoto replied:
Implemented MIN/MAX in show command. I restricted it for Integral types only, for safety

felipepessoto replied:
Changed to support Spark 3.5 show(), which uses ToPrettyString instead of CAST

felipepessoto commented:
Hi folks, did you have a chance to review this?
Thanks

felipepessoto commented:
@scottsand-db

felipepessoto commented:
@vkorukanti, @scottsand-db, do you think we'll be able to complete this before 2.4 release?

felipepessoto commented:
@scottsand-db, @vkorukanti, do we still plan to go ahead with these improvements? Let me know to rebase the changes.

scottsand-db commented:
@felipepessoto - thanks for following up. We are super swamped right now getting a few final features ready for next Delta release  ... we will follow up when we can!

felipepessoto commented:
#1763

henlue commented:
I'm wondering if this is still on the agenda? I think it would be a wonderful enhancement.

There are many practical use cases where performance improvements on such min/max queries would make a difference. Two examples:
 - when incrementally loading data to a table, often the first step is to query the max timestamp of that table in order to figure out from where to continue loading more data
 - BI tools will query the min max values of columns to configure the ranges for their filters or slicers

felipepessoto commented:
We have some folks asking for more improvements using stats here and in other issues/PRs. I think it would help in a couple of scenarios like @henlue mentioned.

https://github.com/delta-io/delta/issues/1192
https://github.com/delta-io/delta/issues/1916
https://github.com/delta-io/delta/pull/1377

@scottsand-db, @vkorukanti, @dennyglee what would be the best way get community feedback about this? Creating a new issue and asking people to thumbs up would be useful? Is it something maintainers use to prioritize the new features?

Thanks

dennyglee commented:
Hey @felipepessoto - good call out, I would probably create new issues for each one so we can garner thumbs up as well as do design reviews (if required), eh?!

vkorukanti commented:
@felipepessoto Sorry for the delay. This is really a great optimization to have. Could you please rebase the changes? I think this will be a great addition to [Delta 3.1](https://github.com/delta-io/delta/milestone/25).

felipepessoto commented:
@vkorukanti I rebased it and adjusted to Spark 3.5 show()

vkorukanti commented:
> @vkorukanti I rebased it and adjusted to Spark 3.5 show()

Thank you @felipepessoto. Bit swamped this week, but will definitely take a look at it next week. thx!

vkorukanti commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -60,37 +127,199 @@ trait OptimizeMetadataOnlyDeltaQuery {
-          // The following code is following how Spark casts a long type to a string type.
-          // See org.apache.spark.sql.catalyst.expressions.Cast#castToString.
-          UTF8String.fromString(count.toString)
+      Nil,
```
comment to indicate that we match only aggs with no group bys

vkorukanti commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -60,37 +127,199 @@ trait OptimizeMetadataOnlyDeltaQuery {
+    DoubleType,
+    DateType)
+
+  case class DeltaColumnStat(
```
nit: single line and a one-line comment

vkorukanti commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -60,37 +127,199 @@ trait OptimizeMetadataOnlyDeltaQuery {
+    min: Any,
+    max: Any)
+
+  def extractGlobalColumnStats(tahoeLogFileIndex: TahoeLogFileIndex):
```
nit: private

vkorukanti replied:
Please also add some docs.

vkorukanti replied:
This is a relatively large method and has too much logic in it. Please split this method into smaller methods each one with clear functionality.

vkorukanti commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -60,37 +127,199 @@ trait OptimizeMetadataOnlyDeltaQuery {
+    Map[String, DeltaColumnStat] = {
+
+      val dataColumns = snapshot.statCollectionPhysicalSchema
+        .filter(col => columnStatsSupportedDataTypes.contains(col.dataType))
```
can we exit early here if the dataColumns is empty?

felipepessoto replied:
We do in the next `if` statement. If dataColumns is empty the lazy variables won't be evaluated

vkorukanti commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -60,37 +127,199 @@ trait OptimizeMetadataOnlyDeltaQuery {
+
+      // DELETE operations creates AddFile records with 0 rows, and no column stats.
+      // We can safely ignore it since there is no data.
+      lazy val files = deltaScanGenerator.filesWithStatsForScan(Nil)
```
we also run `deltaScanGenerator.filesWithStatsForScan` for `count(1)` optimization. Should we share the result in case the query has `count(1), min(col1)`.

felipepessoto replied:
Done

vkorukanti commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -60,37 +127,199 @@ trait OptimizeMetadataOnlyDeltaQuery {
+
+    CaseInsensitiveMap(
+      extractGlobalColumnStatsDeltaLog(snapshot).++
+      (extractGlobalPartitionedColumnStatsDeltaLog(snapshot)))
```
do we need to extract the partition column stats always? Can this be optional based on whether there is agg expr on partition column?

vkorukanti commented on spark/src/test/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuerySuite.scala:
```diff
@@ -188,99 +640,261 @@ class OptimizeMetadataOnlyDeltaQuerySuite
+      s"SELECT (SELECT COUNT(*) FROM $testTableName WHERE id > 0)")
+
+    checkOptimizationIsNotTriggered(
+      s"SELECT (SELECT MAX(id) FROM $testTableName WHERE id > 0)")
```
test case when the filter is both on data and partition columns?

vkorukanti replied:
Also is there is a case where the filter is just on partition columns?

felipepessoto replied:
Added

vkorukanti commented on spark/src/test/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuerySuite.scala:
```diff
@@ -188,99 +640,261 @@ class OptimizeMetadataOnlyDeltaQuerySuite
-      Seq(Row(totalNonNullData)))
+  test("count-min-max - not supported - join") {
+    checkOptimizationIsNotTriggered(
+      s"SELECT COUNT(*) FROM $testTableName A, $testTableName B")
```
Looks like all these not supported queries can be in a Seq and just create a test for each one of them. Reduces the number of lines in the code and improves readability to see what tests are convered.

felipepessoto replied:
Done

vkorukanti reviewed (changes requested):
Did one review round. Looking great!

vkorukanti commented on spark/src/test/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuerySuite.scala:
```diff
@@ -17,141 +17,594 @@
   }
 
-  test("Select Count: table alias") {
+  test("min-max - simple query") {
```
is it possible to create a test matrix? that way we can see what is covered and what is not. There are too many cases.

felipepessoto replied:
Done

felipepessoto commented:
Thanks Venki. I’m out of town, I’m probably working on it next week.

felipepessoto commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -60,37 +127,199 @@ trait OptimizeMetadataOnlyDeltaQuery {
+      // when we create the Local Relation, example
+      // SELECT MAX(Column2) FROM (SELECT Column1 AS Column2 FROM TableName)
+      // the AggregateExpression contains a reference to Column2, instead of Column1
+      PhysicalOperation(fields, Nil, DeltaTable(i: TahoeLogFileIndex)))
```
Not sure if I understand correctly. Renamed TahoeLogFileIndex param to fileIndex

felipepessoto commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -60,37 +127,199 @@ trait OptimizeMetadataOnlyDeltaQuery {
-  object ShowCountStarDeltaTable {
-    def unapply(plan: Aggregate): Option[UTF8String] = plan match {
+
+  object AggregateDeltaTable {
```
Done

felipepessoto commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -60,37 +127,199 @@ trait OptimizeMetadataOnlyDeltaQuery {
-          PhysicalOperation(_, Nil, DeltaTable(i: TahoeLogFileIndex))) if i.partitionFilters.isEmpty
-        => extractGlobalCount(i)
-      case _ => None
+  val columnStatsSupportedDataTypes: HashSet[DataType] = HashSet(
```
Created it in AggregateDeltaTable

felipepessoto commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -60,37 +127,199 @@ trait OptimizeMetadataOnlyDeltaQuery {
+            Count(Seq(Literal(1, _))) | Min(_) | Max(_), Complete, false, None, _), _) => true
+          case Alias(ToPrettyString(AggregateExpression(
+            Count(Seq(Literal(1, _))) | Min(_) | Max(_), Complete, false, None, _), _), _) => true
+          case _ => false
```
Done, and added the data type check.

felipepessoto commented:
I did the first refactor to address some of the comments. I resolved the ones I fixed.

felipepessoto commented on spark/src/test/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuerySuite.scala:
```diff
@@ -17,141 +17,594 @@
+  }
+
+  test("min-max - column mapping enabled") {
+    val tableName = "TestColumnMapping"
```
Thanks. Done.

felipepessoto commented:
I addressed all the comments, except split the "extract from stats" method, which could have some trade-off.

Test is also failing with this error:

`/home/runner/work/delta/delta/spark/src/test/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuerySuite.scala: illegal start of simple expression: Token(RPAREN,),4940,))`

It doesn't happen locally. No idea what it is as it doesn't show error line. Let me know if you have any clue.

Thanks.

vkorukanti commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -17,80 +17,352 @@
+      .map(x => x._1).toSet
+
+    // Creates a tuple with physical name to avoid recalculating it multiple times
+    val dataColumnsWithStats = dataColumns.map(x => (x, DeltaColumnMapping.getPhysicalName(x)))
```
one suggestion to simplify the code:
1) Add a utility method to get the `Column` ref for min/max/nullCount/numRecords for regular or partition columns from the Dataframe `deltaScanGenerator.filesWithStatsForScan`. It abstracts out the physical name conversion and the lookup for partition or data column. If the column is a partition column, then it also takes care of type-casting the string partition value to the appropriate data type value.
2) The next step is to construct the expression using these refs that validate the stats and then return the min and max. the existing expression you have should work.

vkorukanti commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -17,80 +17,352 @@
+    val statsResults = files.select(columnsToQuery: _*).selectExpr(minMaxExpr: _*).head
+
+    (Some(numRecords), dataColumnsWithStats
+      .filter(x => statsResults.getAs[Boolean](s"complete_${x._2}"))
```
If stats are not complete for one of the requested columns, should we just return the empty stats? Can this rule use stats available for partial set of columns referenced in the agg expressions?

felipepessoto replied:
Currently it can't use partial set of columns. I'm not doing the extra check here to avoid running the query twice for success case

vkorukanti commented on spark/src/test/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuerySuite.scala:
```diff
@@ -188,99 +644,249 @@ class OptimizeMetadataOnlyDeltaQuerySuite
+  // Tests to validate the optimizer won't incorrectly change queries it can't correctly handle
+
+  Seq((s"SELECT COUNT(*) FROM $mixedStatsTableName", "missing stats"),
+    (s"SELECT COUNT(*) FROM $noStatsTableName", "missing stats"),
```
few tests to add (apologies if they are already present)
1) table with DVs
2) empty table
3) table with few AddFiles having zero rows

felipepessoto replied:
Done.
We have an existing test for count + empty table. I added the other cases.
For min/max + empty table, that case is not supported. (theoretically it could be supported, but I preferred to be safe and don't optimize, it won't read any parquet anyway)

lzlfred commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -17,80 +17,352 @@
+import java.util.Locale
+
+/** Optimize COUNT, MIN and MAX expressions on Delta tables.
+ * This optimization is only applied when the following conditions are met:
```
also add the condition that all AddFiles in the table must have stats on columns of interest ? (obvious but just for completeness)

felipepessoto replied:
Done

lzlfred commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -17,80 +17,352 @@
+      val dataType = columnAndPhysicalName._1.dataType
+      val physicalName = columnAndPhysicalName._2
+
+      Seq(col(s"stats.minValues.`$physicalName`").cast(dataType).as(s"min.$physicalName"),
```
nit: make minValues/maxValues/nullCount const string and reuse all across.

felipepessoto replied:
Done

lzlfred reviewed (commented):
LGTM. minor comments.

felipepessoto commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -17,80 +17,352 @@
+      // from stats when we create the Local Relation, example
+      // SELECT MAX(Column2) FROM (SELECT Column1 AS Column2 FROM TableName)
+      // the AggregateExpression contains a reference to Column2, instead of Column1
+      case _: AttributeReference => true
```
Yes, currently this optimization doesn't support nested columns. We have a test: "optimization not supported - nested columns"

felipepessoto commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -17,80 +17,352 @@
+    lazy val numFiles: Long = filesStatsCount.getAs[Long]("countNonEmptyFiles")
+
+    val dataColumns = snapshot.statCollectionPhysicalSchema.filter(col =>
+      AggregateDeltaTable.isSupportedDataType(col.dataType) &&
```
Right, we don't need it anymore. Previously this method was getting all the columns stats. I'll remove it.

`lowerCaseColumnNames` condition makes sense. The only thing is I have to change `extractCountMinMaxFromDeltaLog` to first get partition values, which is reasonable as it returns `Map.empty` if no partition column is found in `snapshot.metadata.partitionSchema`

felipepessoto commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -17,80 +17,352 @@
+
+/** Optimize COUNT, MIN and MAX expressions on Delta tables.
+ * This optimization is only applied when the following conditions are met:
+ * - The MIN/MAX columns data type is supported by the optimization (ByteType, ShortType,
```
Done

felipepessoto commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -17,80 +17,352 @@
-  private def extractGlobalCount(tahoeLogFileIndex: TahoeLogFileIndex): Option[Long] = {
-    // account for deleted rows according to deletion vectors
+  private def createLocalRelationPlan(
+    plan: Aggregate,
```
Done

felipepessoto commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -17,80 +17,352 @@
+  case class DeltaColumnStat(min: Any, max: Any)
+
+  private def extractCountMinMaxFromStats(
+    deltaScanGenerator: DeltaScanGenerator,
```
Done

felipepessoto commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -17,80 +17,352 @@
+    def unapply(plan: Aggregate): Option[TahoeLogFileIndex] = plan match {
+      case Aggregate(
+        Nil, // GROUP BY not supported
+        aggExpr: Seq[Alias @unchecked], // Underlying type is not checked because of type erasure.
```
Done

weiluo-db commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -17,80 +17,355 @@
+ * - COUNT has no DISTINCT.
+ * - Query has no filters.
+ * - Query has no GROUP BY.
+ * Example of valid query: SELECT COUNT(*), MIN(id), MAX(id) FROM MyDeltaTable
```
Nit: Perhaps s/MAX(id)/MAX(partition_col), just to highlight that partition columns are stats-optimizable.

weiluo-db commented on spark/src/test/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuerySuite.scala:
```diff
@@ -17,141 +17,600 @@
+      dfPart2.write.format("delta").mode(SaveMode.Append).saveAsTable(mixedStatsTableName)
+
+      // Run updates to generate more Delta Log and trigger a checkpoint
+      // and make sure stats works after checkpoints
```
Given you are covering checkpoints already, how about when `delta.checkpoint.writeStatsAsStruct=true`: https://docs.delta.io/latest/table-properties.html?

felipepessoto replied:
We rely on filesWithStatsForScan, so it shouldn't make difference. Currently writeStatsAsStruct is not working: https://github.com/delta-io/delta/issues/1719. When that issue is fixed, my expectation is that filesWithStatsForScan will handle it.

felipepessoto commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -17,80 +17,355 @@
+        Seq(InternalRow.fromSeq(rewrittenAggregationValues)))
+      r
+    } else {
+      plan
```
Done. Just for my information, when running unit tests, where are the logs saved?

felipepessoto replied:
Correct, the method createLocalRelationPlan is only called if query matches AggregateDeltaTable, where we check these conditions.

The test `optimization not supported - filter - SELECT COUNT(*) FROM table_basic WHERE id > 0` don't generate logs, while `optimization not supported - missing stats - SELECT COUNT(*) FROM  table_mixstats` does

weiluo-db commented on spark/src/main/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuery.scala:
```diff
@@ -16,81 +16,358 @@
+    (rowCount, CaseInsensitiveMap(columnStats ++ partitionedValues))
+  }
+
+  object AggregateDeltaTable {
```
Nit: How about s/AggregateDeltaTable/MetadataOptimizableAggregate? The name may be a bit verbose, but it makes the object's intention very clear.

felipepessoto replied:
Correct, the method createLocalRelationPlan is only called if query matches AggregateDeltaTable, where we check these conditions

weiluo-db reviewed (approved):
LGTM (pending @vkorukanti 's final pass)!

vkorukanti commented on spark/src/test/scala/org/apache/spark/sql/delta/perf/OptimizeMetadataOnlyDeltaQuerySuite.scala:
```diff
@@ -17,141 +17,600 @@
+      spark.sql(s"DELETE FROM $testTableName WHERE id = 11")
     }
+
+    val result = spark.sql(s"SELECT COUNT(*), MIN(id), MAX(id) FROM $testTableName").head
```
@felipepessoto Can you make a change to collect these golden results without the optimization enabled? The optimization is enabled by default.

felipepessoto replied:
Done

vkorukanti reviewed (commented):
lgtm pending one comment.

vkorukanti reviewed (approved):
lgtm

Thank you for contributing this optimizaiton.

felipepessoto commented:
Flink tests are flaky? Previous build succeeded
